Simplified VC calculations

kpentland opened the pull request:
In this PR, I've simplified the VC calculation process.

Primarily, this meant simplifying the way that plasma shape parameters are defined and used within the VC framework. This required some API changes and the Example09 notebook has been updated to reflect these changes. Hopefully they will come in handy.

georgeholt1 commented:
Thanks for this @kpentland! Definitely a lot simpler now.

I have a couple of questions/points. Some aren't related to what you've changed but I think they're still worth raising at this point.

- In `apply_VC`:
   https://github.com/FusionComputingLab/freegsnke/blob/2dff80ad9e040eb11809cc74b967fcc3bda77fb4/freegsnke/virtual_circuits.py#L529-L531

  If `self.target_calculator` exists, might it be outdated, e.g. from a different VC calculation using the same object?

  Possible fix:

  ```python
  old_target_values = VC_object.target_calculator(eq)
  # ... later ...
  new_target_values = VC_object.target_calculator(eq_new)
  ```
- In `prepare_build_dIydI_j`:

  https://github.com/FusionComputingLab/freegsnke/blob/2dff80ad9e040eb11809cc74b967fcc3bda77fb4/freegsnke/virtual_circuits.py#L245-L249

  Can the change in plasma current ever be zero here? If so, could be a division by zero error.

- `self.solver` is called quite a few times, but the `solver` attribute is only set when `define_solver` is called. I added checks for an attribute error to each of these.

- I think there's either a bug in `calculate_VC` or the documentation is wrong: the `starting_dI` argument is specified as a float in the docstring, but then is indexed in a couple of places as if it were an array:
  https://github.com/FusionComputingLab/freegsnke/blob/2dff80ad9e040eb11809cc74b967fcc3bda77fb4/freegsnke/virtual_circuits.py#L421

  https://github.com/FusionComputingLab/freegsnke/blob/2dff80ad9e040eb11809cc74b967fcc3bda77fb4/freegsnke/virtual_circuits.py#L423

  Based on the default behaviour (when `None` is passed), I assume this is always supposed to be an array?

kpentland commented:
Thanks for the feedback @georgeholt1!

1. First point: agreed, not sure why that check was in there so I've removed and used the following as suggested:

`old_target_values = VC_object.target_calculator(eq)`
`new_target_values = VC_object.target_calculator(eq_new)`


2. It's very unlikely either of those two will ever be zero. For the first `self._nIy` to be zero, it would require your input equilibrium to have zero plasma current (which would be odd...). And for the second, `rel_ndIy_0` to be zero, it means there is no change in the plasma current density between the initial equilibrium and the perturbed one. I've added a catch in there just in case, if it's triggered the user needs to increase the initial step size in the currents. 

3. Nice spot, thanks for that. 

4. Ah yes, the doc strings are wrong there thanks. I've updated and it should be an array. I've also changed the print statements for these outputs as they're a bit confusing.
